### PR TITLE
admin user

### PR DESCRIPTION
**Pull request by RA341**
## Proposed changes
--------------------
Implements #168, but instead of adding the user in the config file, the first account created automatically becomes the admin

## Types of changes
--------------------
What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist
--------------------
_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. This is simply a reminder of what we are going to look for before merging your code._

- [x] My changeset covers only what is described above (no extraneous changes)
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: devU-api/src/entities/user/user.controller.ts [hunk 1]
```diff
@@ -48,6 +49,48 @@ export async function getByCourse(req: Request, res: Response, next: NextFunctio
...
+  } catch (e) {
+    next(e)
+  }
+}
+
+
+// delete an admin, only an admin can delete an admin
+export async function deleteAdmin(req: Request, res: Response, next: NextFunction) {
+  try {
+    let newAdminUserId = req.body.newAdminUserId
```
Comment (NicholasMy) [inline devU-api/src/entities/user/user.controller.ts:71]: Use a more appropriate variable name

File: devU-api/src/entities/user/user.controller.ts [hunk 2]
```diff
@@ -48,6 +49,48 @@ export async function getByCourse(req: Request, res: Response, next: NextFunctio
...
+    if (!newAdminUserId) {
+      return res.status(404).send('Not found')
+    }
+    await UserService.softDeleteAdmin(newAdminUserId)
+    res.status(204).send('User is no longer admin')
```
Comment (NicholasMy) [inline devU-api/src/entities/user/user.controller.ts:76]: A 204 shouldn't have a body

File: devU-api/src/entities/user/user.middlware.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+// is admin middleware, use this when marking an endpoint as only accessible by admin
+// different from userCourse permissions. this is attached to a user instead of a course level permission
+export async function isAdmin(req: Request, res: Response, next: NextFunction) {
+  const userId = req.currentUser?.userId
+  if (!userId) {
+    return res.status(403).send('Unauthorized')
+  }
+
+  const isAdmin = await UserService.isAdmin(userId)
+  if (!isAdmin!.isAdmin!) return res.status(403).send('unauthorized')
```
Comment (NicholasMy) [inline devU-api/src/entities/user/user.middlware.ts:13]: Try to be consistent with capitalization on "Unauthorized"

File: devU-api/src/entities/user/user.service.ts [hunk 4]
```diff
@@ -29,6 +36,35 @@ export async function retrieve(id: number) {
...
+
+export async function createAdmin(id: number) {
+  return await connect().update(id, { isAdmin: true })
+}
+
+// soft deletes an admin
+export async function softDeleteAdmin(id: number) {
+  let res = await connect().count({ take: 2, where: { isAdmin: true } })
+  // check if this deletes the last admin
+  // there must always be at least 1 admin
```
Comment (NicholasMy) [inline devU-api/src/entities/user/user.service.ts:54]: Good thinking!

**Review by NicholasMy — CHANGES_REQUESTED** [review]

**Comment by RA341** [comment]
fixed it

**Pull request merged**